cmd/snap-confine/ns-support.c

```diff
@@ -253,32 +322,57 @@ void sc_create_or_join_ns_group(struct sc_ns_group *group,
 	// sc_discard_preserved_ns_group() it will revert to a regular file.  If
 	// snap-confine is killed for whatever reason after the file is created but
 	// before the file is bind-mounted it will also be a regular file.
-	mnt_fd =
-	    openat(group->dir_fd, mnt_fname,
-		   O_CREAT | O_RDONLY | O_CLOEXEC | O_NOFOLLOW, 0600);
+	mnt_fd = openat(group->dir_fd, mnt_fname,
+			O_CREAT | O_RDONLY | O_CLOEXEC | O_NOFOLLOW, 0600);
 	if (mnt_fd < 0) {
 		die("cannot open mount namespace file for namespace group %s",
 		    group->name);
 	}
-	// Check if we got an nsfs-based file or a regular file. This can be
-	// reliably tested because nsfs has an unique filesystem type NSFS_MAGIC.
-	// On older kernels that don't support nsfs yet we can look for
-	// PROC_SUPER_MAGIC instead.
+	// Check if we got an nsfs-based or procfs file or a regular file. This can
+	// be reliably tested because nsfs has an unique filesystem type
+	// NSFS_MAGIC.  On older kernels that don't support nsfs yet we can look
+	// for PROC_SUPER_MAGIC instead. 
 	// We can just ensure that this is the case thanks to fstatfs.
-	struct statfs buf;
-	if (fstatfs(mnt_fd, &buf) < 0) {
-		die("cannot perform fstatfs() on an mount namespace file descriptor");
+	struct statfs ns_statfs_buf;
+	if (fstatfs(mnt_fd, &ns_statfs_buf) < 0) {
+		die("cannot perform fstatfs() on the mount namespace file descriptor");
+	}
+	// Stat the mount namespace as well, this is later used to check if the
+	// namespace is used by other processes if we are considering discarding a
+	// stale namespace.
+	struct stat ns_stat_buf;
+	if (fstat(mnt_fd, &ns_stat_buf) < 0) {
+		die("cannot perform fstat() on the mount namespace file descriptor");
 	}
 #ifndef NSFS_MAGIC
 // Account for kernel headers old enough to not know about NSFS_MAGIC.
 #define NSFS_MAGIC 0x6e736673
 #endif
-	if (buf.f_type == NSFS_MAGIC || buf.f_type == PROC_SUPER_MAGIC) {
+	if (ns_statfs_buf.f_type == NSFS_MAGIC
+	    || ns_statfs_buf.f_type == PROC_SUPER_MAGIC) {
+		char fname[PATH_MAX] = { 0, };
+		char base_snap_rev[PATH_MAX] = { 0, };
```

Comment:
Why the syntax initializing these to a zero first entry here?  We don't do that in the change above. What's the overall convention for snap-confine, and what's the rationale behind it?

Reply:
This is a new thing that @jdstrand has requested that we do. I'll do a pass to ensure all the buffers are zero-initialized so that we have consistency.